src/basicProof/BasicProvers.sml

```diff
@@ -285,6 +295,188 @@ fun Induct (g as (_,w)) =
  end
  handle e => raise wrap_exn "BasicProvers" "Induct" e
 
+(*---------------------------------------------------------------------------*)
+(* No loss of generality                                                     *)
+(*---------------------------------------------------------------------------*)
+
+fun LIST_SPLICE l =
+  case l of
+  [] => raise ERR "LIST_SPLICE" "Empty list"
+  | [t1] => REFL t1
+  | [t1, t2] => QCONV (SPLICE_CONJ_CONV) (mk_conj (t1, t2))
+  | (t::l) =>
+    let
+      val deep_thm = LIST_SPLICE l
+      (* If “deep_thm” is «l ⇔ r», “lifted_deep_thm” is «t ∧ l ⇔ t ∧ r». UOK *)
+      val lifted_deep_thm = AP_TERM (mk_comb (conjunction, t)) deep_thm
+    in
+      RIGHT_CONV_RULE SPLICE_CONJ_CONV lifted_deep_thm
+    end
+
+fun DISCH_CONJ l thm =
+  let
+    fun loop l conj_thm thm =
+      case l of
+      [x] =>
+        PROVE_HYP conj_thm thm
+      | [] =>
+        thm
+      | (x::rest) =>
+        loop rest (CONJUNCT2 conj_thm) (PROVE_HYP (CONJUNCT1 conj_thm) thm)
+  in
+    case l of
+      [] => thm
+    | [x] => DISCH x thm
+    | _ => let val t = list_mk_conj l in DISCH t (loop l (ASSUME t) thm) end
+  end
+
+fun UNDISCH_CONJ l thm =
+  let
+    fun fold_rator (t, acc) =
+      CONJ (ASSUME t) acc
+  in
+    case l of
+    [] => thm
+    | [t] => MP thm (ASSUME t)
+    | _ =>
+      let
+        val l = rev l
+      in
+        MP thm (foldl fold_rator (ASSUME (hd l)) (tl l))
+      end
+  end
+  handle HOL_ERR _ => raise ERR "UNDISCH_CONJ" ""
+
+(* Proves
+  ⊢ (¬P ⇒ (∀(vars). P_hyp ⇒ t) ⇒ t) ⇒   (* “wlog_thm”. *)                  UOK
+     (∀(vars). P ∧ hyp ⇒ t) ⇒           (* “new_sg_thm”. *)                UOK
+     hyp ⇒                                                                 UOK
+     t»                                                                    UOK
+except that when "hyp" is "NONE", it is omitted the obvious way. If "hyp" is
+none then "p_hyp" is just "hyp" else it is the splice of "p" with "hyp". The
+tuple returned is the aforementioned theorem and "P_hyp". *)
+fun wlog_rule vars p hyp t =
+  let
+    val the = liteLib.the
+    fun forall vars t = list_mk_forall (vars, t)
+    fun implies t1 t2 = mk_imp (t1, t2)
+    val new_sg_t = forall vars (implies (case hyp of
+                                           SOME hyp_t => mk_conj (p, hyp_t)
+                                         | NONE       => p)
+                                        t)
+    val new_sg_thm = ASSUME new_sg_t
+    val thm_p = MP (SPECL vars new_sg_thm)
+                   (case hyp of
+                      SOME hyp_t => CONJ (ASSUME p) (ASSUME hyp_t)
+                    | NONE       => (ASSUME p))
+    val wlog_t = ref NONE
+    val wlog_ant = ref NONE
+    val splice_thm = ref NONE
+    val _ =
```

Comment:
I see no reason to use references here. Please use pure code for deriving `wlog_and` and `splice_thm`.